src/transaction_history.rs

```diff
@@ -0,0 +1,500 @@
+    let tx = StoredRichTx::new(id, action, coins, memo, block);
+
+    if minter != recipient {
+        append_tx(store, &tx, recipient)?;
+    }
+    append_tx(store, &tx, minter)?;
+
+    Ok(())
+}
+
+pub fn store_burn<S: Storage>(
+    store: &mut S,
+    owner: &CanonicalAddr,
+    burner: &CanonicalAddr,
+    amount: Uint128,
+    denom: String,
+    memo: Option<String>,
+    block: &cosmwasm_std::BlockInfo,
+) -> StdResult<()> {
+    let id = increment_tx_count(store)?;
+    let coins = Coin { denom, amount };
+    let action = StoredTxAction::burn(owner.clone(), burner.clone());
+    let tx = StoredRichTx::new(id, action, coins, memo, block);
+
+    if burner != owner {
+        append_tx(store, &tx, owner)?;
+    }
+    append_tx(store, &tx, burner)?;
+
+    Ok(())
+}
+
+pub fn store_deposit<S: Storage>(
+    store: &mut S,
+    recipient: &CanonicalAddr,
+    amount: Uint128,
+    denom: String,
+    block: &cosmwasm_std::BlockInfo,
+) -> StdResult<()> {
+    let id = increment_tx_count(store)?;
+    let coins = Coin { denom, amount };
+    let action = StoredTxAction::deposit();
+    let tx = StoredRichTx::new(id, action, coins, None, block);
+
+    append_tx(store, &tx, recipient)?;
+
+    Ok(())
+}
+
+pub fn store_redeem<S: Storage>(
+    store: &mut S,
+    redeemer: &CanonicalAddr,
+    amount: Uint128,
+    denom: String,
+    block: &cosmwasm_std::BlockInfo,
+) -> StdResult<()> {
+    let id = increment_tx_count(store)?;
+    let coins = Coin { denom, amount };
+    let action = StoredTxAction::redeem();
+    let tx = StoredRichTx::new(id, action, coins, None, block);
+
+    append_tx(store, &tx, redeemer)?;
+
+    Ok(())
+}
+
+fn append_tx<S: Storage>(
+    store: &mut S,
+    tx: &StoredRichTx,
+    for_address: &CanonicalAddr,
+) -> StdResult<()> {
+    let mut store = PrefixedStorage::multilevel(&[PREFIX_TXS, for_address.as_slice()], store);
+    let mut store = AppendStoreMut::attach_or_create(&mut store)?;
+    store.push(tx)
+}
+
+fn append_transfer<S: Storage>(
+    store: &mut S,
+    tx: &StoredLegacyTransfer,
+    for_address: &CanonicalAddr,
+) -> StdResult<()> {
+    let mut store = PrefixedStorage::multilevel(&[PREFIX_TRANSFERS, for_address.as_slice()], store);
+    let mut store = AppendStoreMut::attach_or_create(&mut store)?;
+    store.push(tx)
+}
+
+pub fn get_txs<A: Api, S: ReadonlyStorage>(
+    api: &A,
+    storage: &S,
+    for_address: &CanonicalAddr,
+    page: u32,
+    page_size: u32,
+) -> StdResult<(Vec<RichTx>, u64)> {
+    let store = ReadonlyPrefixedStorage::multilevel(&[PREFIX_TXS, for_address.as_slice()], storage);
+
+    // Try to access the storage of txs for the account.
+    // If it doesn't exist yet, return an empty list of transfers.
+    let store = AppendStore::<StoredRichTx, _, _>::attach(&store);
+    let store = if let Some(result) = store {
+        result?
+    } else {
+        return Ok((vec![], 0));
+    };
```

Comment:
(I guess it's from the old reference impl. but) I think you can replace it with something like `.unwrap_or()`

Reply:
tried but there wasn't anything i could find that did the right thing here

Reply:
You're right, I haven't realized that the return statement was in the scope of that whole function